### PR TITLE
feat: convert regmapper to chisel3

### PR DESCRIPTION
**Pull request by tianrui-wei**
Converts the regmapper directory to chisel3

Signed-off-by: Tianrui Wei <tianrui@tianruiwei.com>

**Type of change**: other enhancement

**Impact**: no functional change

**Development Phase**: implementation

**Release Notes**

### CODE REVIEW

File: src/main/scala/regmapper/RegField.scala [hunk 1]
```diff
@@ -2,8 +2,8 @@
 
 package freechips.rocketchip.regmapper
 
-import Chisel._
-import chisel3.util.{ReadyValidIO}
+import chisel3._
+import chisel3.util._
```
Comment (sequencer) [inline src/main/scala/regmapper/RegField.scala:6]: import specific utils from `util`

File: src/main/scala/regmapper/RegField.scala [hunk 2]
```diff
@@ -125,7 +125,7 @@ object RegField
   // and to clear bits when the bus writes bits of value 1.
   // Setting takes priority over clearing.
   def w1ToClear(n: Int, reg: UInt, set: UInt, desc: Option[RegFieldDesc] = None): RegField =
-    RegField(n, reg, RegWriteFn((valid, data) => { reg := ~(~reg | Mux(valid, data, UInt(0))) | set; Bool(true) }),
+    RegField(n, reg, RegWriteFn((valid, data) => { reg := (~((~reg).asUInt | Mux(valid, data, 0.U))).asUInt | set; true.B }),
```
Comment (sequencer) [inline src/main/scala/regmapper/RegField.scala:128]: I don't think we should add `asUInt` here, that's the false error from IDEA.

File: src/main/scala/regmapper/RegMapper.scala [hunk 3]
```diff
@@ -2,12 +2,12 @@
 
 package freechips.rocketchip.regmapper
 
-import Chisel._
-
+import chisel3._
+import chisel3.internal.sourceinfo.SourceInfo
+import chisel3.util._
```
Comment (sequencer) [inline src/main/scala/regmapper/RegMapper.scala:7]: same here.

File: src/main/scala/regmapper/RegMapper.scala [hunk 4]
```diff
@@ -76,28 +76,28 @@ object RegMapper
...
     val regSize = 1 << maskBits
     def regIndexI(x: Int) = ofBits((maskFilter zip toBits(x)).filter(_._1).map(_._2))
-    def regIndexU(x: UInt) = if (maskBits == 0) UInt(0) else
+    def regIndexU(x: UInt) = if (maskBits == 0) 0.U else
       Cat((maskFilter zip x.asBools).filter(_._1).map(_._2).reverse)
 
-    val findex = front.bits.index & maskMatch
-    val bindex = back .bits.index & maskMatch
+    val findex = (front.bits.index & maskMatch.asUInt).asUInt
+    val bindex = (back .bits.index & maskMatch.asUInt).asUInt
```
Comment (sequencer) [inline src/main/scala/regmapper/RegMapper.scala:90]: don't need `asUInt` here, if you wanna make IDEA happy, try this:
```suggestion
    val findex: UInt = (front.bits.index & maskMatch)
    val bindex: UInt = (back .bits.index & maskMatch)
```

File: src/main/scala/regmapper/RegisterCrossing.scala [hunk 5]
```diff
@@ -2,25 +2,25 @@
 
 package freechips.rocketchip.regmapper
 
-import Chisel._
-import chisel3.util.{Irrevocable}
+import chisel3._
+import chisel3.util._
```
Comment (sequencer) [inline src/main/scala/regmapper/RegisterCrossing.scala:6]: same here.

File: src/main/scala/regmapper/RegisterCrossing.scala [hunk 6]
```diff
@@ -38,19 +38,19 @@ class BusyRegisterCrossing extends Module {
 
 class RegisterCrossingAssertion extends Module {
   val io = new Bundle {
```
Comment (sequencer) [inline src/main/scala/regmapper/RegisterCrossing.scala:40]: add IO wrapper

File: src/main/scala/regmapper/Test.scala [hunk 7]
```diff
@@ -21,40 +22,40 @@ object LFSR16Seed
...
     LFSR16Seed(seed)(0)
   }
 
   def delay(x: Int): Bool => Bool = { ready =>
-    val reg = RegInit(UInt(0, width = log2Ceil(x+1)))
-    val valid = reg === UInt(0)
-    reg := Mux(ready && valid, UInt(x), Mux(valid, UInt(0), reg - UInt(1)))
+    val reg = RegInit(0.U(log2Ceil(x+1).W))
+    val valid = reg === 0.U
+    reg := Mux(ready && valid, (x).U, Mux(valid, 0.U, reg - 1.U))
```
Comment (sequencer) [inline src/main/scala/regmapper/Test.scala:58]:
```suggestion
    reg := Mux(ready && valid, x.U, Mux(valid, 0.U, reg - 1.U))
```

File: src/main/scala/regmapper/Test.scala [hunk 8]
```diff
@@ -71,21 +72,21 @@ class RRTestRequest(val bits: Int,
   wflow: (Bool, Bool, UInt) => (Bool, Bool, UInt)) extends Module
 {
   val io = new Bundle {
```
Comment (sequencer) [inline src/main/scala/regmapper/Test.scala:74]: add IO wrapper

File: src/main/scala/regmapper/Test.scala [hunk 9]
```diff
@@ -21,40 +22,40 @@ object LFSR16Seed
 class RRTestCombinational(val bits: Int, rvalid: Bool => Bool, wready: Bool => Bool) extends Module
 {
   val io = new Bundle {
```
Comment (sequencer) [inline src/main/scala/regmapper/Test.scala:24]: add IO wrapper

File: src/main/scala/regmapper/RegisterCrossing.scala [hunk 10]
```diff
@@ -2,25 +2,25 @@
...
-import Chisel._
-import chisel3.util.{Irrevocable}
+import chisel3._
+import chisel3.util._
 
 import freechips.rocketchip.util.{AsyncQueue,AsyncQueueParams,AsyncResetRegVec}
 
 // A very simple flow control state machine, run in the specified clock domain
 class BusyRegisterCrossing extends Module {
   val io = new Bundle {
```
Comment (sequencer) [inline src/main/scala/regmapper/RegisterCrossing.scala:12]: add IO wrapper

File: src/main/scala/regmapper/RegisterCrossing.scala [hunk 11]
```diff
@@ -77,16 +77,16 @@ class RegisterWriteIO[T <: Data](gen: T) extends Bundle {
...
-  val slave_reset    = Bool(INPUT)
-  val slave_register = gen.asOutput
-  val slave_valid    = Bool(OUTPUT) // is high on 1st cycle slave_register has a new value
+  val slave_clock    = Input(Clock())
+  val slave_reset    = Input(Bool())
+  val slave_register = Output(gen)
+  val slave_valid    = Output(Bool()) // is high on 1st cycle slave_register has a new value
 }
 
 class RegisterWriteCrossing[T <: Data](gen: T, sync: Int = 3) extends Module {
```
Comment (sequencer) [inline src/main/scala/regmapper/RegisterCrossing.scala:92]: add IO wrapper

File: src/main/scala/regmapper/RegisterCrossing.scala [hunk 12]
```diff
@@ -125,22 +125,22 @@ class RegisterWriteCrossing[T <: Data](gen: T, sync: Int = 3) extends Module {
...
   // Slave clock domain
-  val slave_clock    = Clock(INPUT)
-  val slave_reset    = Bool(INPUT)
-  val slave_register = gen.asInput
+  val slave_clock    = Input(Clock())
+  val slave_reset    = Input(Bool())
+  val slave_register = Input(gen)
 }
 
 class RegisterReadCrossing[T <: Data](gen: T, sync: Int = 3) extends Module {
```
Comment (sequencer) [inline src/main/scala/regmapper/RegisterCrossing.scala:146]: add IO wrapper

**Comment by tianrui-wei** [comment]
Thanks for reviewing. The issues mentioned above are addressed

File: src/main/scala/regmapper/RegField.scala [hunk 13]
```diff
@@ -147,16 +147,16 @@ object RegField
     val numFullBytes = reg.getWidth/8
     val numPartialBytes  = if ((reg.getWidth % 8) > 0) 1 else 0
     val numPadBytes = numBytes - numFullBytes - numPartialBytes
-    val pad = reg | UInt(0, width = 8*numBytes)
-    val oldBytes = Vec.tabulate(numBytes) { i => pad(8*(i+1)-1, 8*i) }
-    val newBytes = Wire(init = oldBytes)
-    val valids = Wire(init = Vec.fill(numBytes) { Bool(false) })
+    val pad = reg | 0.U( (8*numBytes).W)
```
Comment (jerryz123) [inline src/main/scala/regmapper/RegField.scala:150]:
```suggestion
    val pad = reg | 0.U((8*numBytes).W)
```

File: src/main/scala/regmapper/RegisterCrossing.scala [hunk 14]
```diff
@@ -216,7 +216,7 @@ object AsyncRWSlaveRegField {
     async_slave_reg.io.en := wr_crossing.io.slave_valid
     async_slave_reg.io.d  := wr_crossing.io.slave_register
 
-    val rd_crossing = Module (new RegisterReadCrossing(UInt(width = width )))
+    val rd_crossing = Module (new RegisterReadCrossing(UInt( width.W )))
```
Comment (jerryz123) [inline src/main/scala/regmapper/RegisterCrossing.scala:219]:
```suggestion
    val rd_crossing = Module (new RegisterReadCrossing(UInt(width.W)))
```

**Review by jerryz123 — APPROVED** [review]

File: src/main/scala/regmapper/RegField.scala (hunk 13 above)
Comment (tianrui-wei) [reply src/main/scala/regmapper/RegField.scala]: Thanks for reviewing! I addressed these changes with the new commit

**Comment by tianrui-wei** [comment]
Thanks for reviewing! I addressed these changes with the new commit

**Review by sequencer — APPROVED** [review]

**Pull request merged**